### PR TITLE
[r] Build against shared library `libtiledbsoma`

### PR DESCRIPTION
**Pull request by johnkerl**
**Issue and/or context:**

* Top-level issue: #1277
* Split-out for Mac/CI issues: #1293
* This PR is a continuation of #1280

**Changes:**

Here is my understanding of how TileDB-SOMA-R install with `libtiledbsoma` works, on Linux and Mac, before and after this PR.

_Before_

All relative to the `apis/r` directory. (Some minor details omitted for brevity.)

* Install R: CI `tools/r-ci.sh bootstrap`
* Install this package's R dependencies: CI `tools/r-ci.sh install_all`
  * That includes the TileDB-R package which will in turn depends on TileDB core
* `copy_source.sh`:
  * Copy `../../libtiledbsoma` to `./src/libtiledbsoma`
* R build/install
  * Invoked by:
    * `R CMD build --no-build-vignettes --no-manual`
    * `R CMD check --no-vignettes --no-manual tiledbsoma_*.tar.gz`
    * `R CMD INSTALL --no-test-load --no-docs --no-html tiledbsoma_*.gz`
  * Callback files into our package include `configure` and `src/Makevars.in`
      * See also https://cran.r-project.org/doc/manuals/R-exts.html for reference
  * Note that we need core with `include/tiledb/tiledb` et al., and `lib/libtiledb.so` or `lib/libtiledb.dylib`
  * Note that we need `libtiledbsoma` with `include/tiledbsoma/tiledbsoma` et al., and `lib/libtiledbsoma.so` or `lib/libtiledbsoma.dylib`
  * Our `configure` is called
    * It uses `pkg-config` to see if the core library exists (nominally in `/usr/local`)
    * If not it uses the `tools/get_tarball.R` script to put core into `./inst/include` and `./inst/lib`
    * It checks for `copy_source.sh` having copied `../../libtiledbsoma` source into `src/libtiledbsoma`
    * It runs `cmake` with `-DTILEDBSOMA_BUILD_R=ON` to install a static `libtiledbsoma.a` into the current directory
    * It runs a `sed` on our `src/Makevars.in` to create `src/Makevars` with the discovered path to the core library
  * Our `src/Makevars` (which is a Makefile) is invoked
    * Its job is to create a `tiledbsoma.so` which is created and managed by the R build infrastructure
    * Its `-I` and `-L` were just computed above
    * It invokes `install_name_tool` to set an `rpath` if the core library is at `inst/tiledb/lib/libtiledb.dylib`, updating the R-infra's `tiledbsoma.so`

_After_

* Install R: same
* Install this package's R dependencies: same
* There is no more `copy_source.sh`
* Instead, from `../..`, `scripts/bld --prefix=/usr/local`
  * This means `libtiledbsoma.so` or `libtiledbsoma.dylib` is now found in `/usr/local/lib` and not relative to the current directory as it was before
* R build/install
  * Note that we still need core with `include/tiledb/tiledb` et al., and `lib/libtiledb.so` or `lib/libtiledb.dylib`
  * Note that we still need `libtiledbsoma` with `include/tiledbsoma/tiledbsoma` et al., and `lib/libtiledbsoma.so` or `lib/libtiledbsoma.dylib`
  * Our `configure` is called
    * It uses `pkg-config` to see if the core library exists (nominally in `/usr/local`)
    * If not it uses the `tools/get_tarball.R` script to put core into `./inst/include` and `./inst/lib`
    * It doesn't do anything about `libtiledbsoma` which was already installed to `/usr/local`
  * Our `src/Makevars` (which is a Makefile) is invoked
    * It invokes `install_name_tool` to set an `rpath` for `/usr/local/lib/libtiledylib.so` (or `.dylib`) updating the R-infra's `tiledbsoma.so`

**Notes for reviewer:**

* The install prefix `/usr/local/` for `libtiledbsoma` is hard-coded `src/Makevars.in` because it is hard-coded in `scripts/bld --prefix=/usr/local`. This could be made programmable in both places.
* Also note what we do in Python:
  * The paths to the core `libtiledb` and the `libtiledbsoma` can, either or both, be passed in to `setup.py`
  * https://github.com/single-cell-data/TileDB-SOMA/blob/1.2.2/apis/python/setup.py#L43-L72

### CODE REVIEW

**Comment by eddelbuettel** [comment]
We can probably turn all other CI yaml files off til the R file behaves.  I have been musing about this but have no good ideas yet ... but does the create libtiledbsoma.so look right on your local machine?   If the link failed it can either have been made wrong (but we trust `cmake`) or is use wrong (but it is a standard R build).  Or maybe the macOS build env?
Very annoying problem and thanks for digging!

**Comment by johnkerl** [comment]
@eddelbuettel I'm just debugging, some outside of CI and some (here) inside of CI. Nothing to announce yet.

I don't get a usable install on my Mac.

**Comment by johnkerl** [comment]
@eddelbuettel CI _does_ need to pass. This is just temporary debug. An issue I'm seeing with #1280 is R Mac CI fails, and R Linux CI gets cancelled. What I'm testing right now is if we skip R Mac CI, will R Linux CI succeed or fail.

**Comment by eddelbuettel** [comment]
@johnkerl It should not surprise you but _I am aware_ as it after first failed in my branch and PR.  I just tried to help, given that you continue to work on very public branches everybody can see.  And comment on.

**Comment by johnkerl** [comment]
@eddelbuettel I understand and appreciate it! Just trying to figure some things out, as I can, and to help out as requested. No clear answers yet ...

**Comment by eddelbuettel** [comment]
I just had to be on my bicycle for a bit to run an errand and the best I idea I had was that we _may_ need an `install_name_tool` statement for the SOMA library just as we do for `tiledb`.  In `src/Makevars{,.in}` we have

```make
	if [ "$(OS)" != "Windows_NT" ] && [ `uname -s` = 'Darwin' ] && [ -f ../inst/tiledb/lib/libtiledb.dylib ] && [ -f tiledbsoma.so ]; then install_name_tool -change libz.1.dylib @rpath/libz.1.dylib ../inst/tiledb/lib/libtiledb.dylib; install_name_tool -add_rpath @loader_path/../tiledb/lib tiledbsoma.so; fi
```

Here the final part is 

```
install_name_tool -add_rpath @loader_path/../tiledb/lib tiledbsoma.so
```

which AFAIK tells the `tiledbsoma.so` (which was just built by R invoking `R CMD COMPILE` and `R CMD SHLIB` behind the scenes for us) to respect the path `../tiledb/lib` as a dynamic linker load path.  _Maybe_ we have to do the same for the location of `libtiledbsoma.so` which (currently) gets copied into `/usr/local/lib`.  

But that is mere speculation on my part. I never use macOS and honestly do not know the ins and outs.  We need a macHead to look at this.

**Comment by johnkerl** [comment]
> But that is mere speculation on my part. I never use macOS and honestly do not know the ins and outs. We need a macHead to look at this.

@eddelbuettel I have a Mac (more than one) & am looking into this yesterday & today as my first (although not only) priority. I'm doing cross-compares between a bare EC2 Linux instance (no tiledb-dev anything), and my personal/non-work Mac which also has no tiledb-dev anything.

I've gotten as much signal out of CI as I'm willing to -- having side-by-side repros on two systems _which I can actually log into_ is proving to be slow, but fruitful.

This PR isn't ready for review -- it's a trial balloon which has served its job -- but progress is being made.

**Comment by eddelbuettel** [comment]
I'll just keep it here because we have to keep it somewhere.  

What may help is the equivalent of 

     ldconfig -p     

which on Linix tells us what the dynamic linker knows, and it knows `libtiledbsoma` (after `sudo ldconfig`):

```
edd@rob:~$ sudo ldconfig -p|grep "tiledb.*\.so"$
        libtiledbsoma.so (libc6,x86-64) => /usr/local/lib/libtiledbsoma.so
        libtiledb.so (libc6,x86-64) => /usr/local/lib/libtiledb.so
edd@rob:~$ 
```

and on the other end to ask the shared library created by R about what it needs:

```
edd@rob:~/git/tiledb-soma/apis/r(de/sc-28200/shared_soma_lib)$ ldd src/tiledbsoma.so | grep tiledb
        libtiledbsoma.so => /usr/local/lib/libtiledbsoma.so (0x00007f69b1279000)
        libtiledb.so.2.16 => /usr/local/lib/libtiledb.so.2.16 (0x00007f69afe00000)
edd@rob:~/git/tiledb-soma/apis/r(de/sc-28200/shared_soma_lib)$ 
```

This happens to match here, so it builds.  We probably need some macOS help to get to the same two steps (?) in order to get to the build.

**Comment by eddelbuettel** [comment]
I see no R build under CI.  Am I looking in the wrong place?

**Comment by johnkerl** [comment]
Solved; re-opening as RFC with @ihnorton

**Comment by eddelbuettel** [comment]
> Solved late last night as noted in this PR's (now-updated) description field; re-opening as RFC with @ihnorton

Sorry, I am dense. What is solved, and where, and by which change?

**Comment by johnkerl** [comment]
@eddelbuettel what is solved is that this now installs on Mac

The reason for altering the existing line is detailed in the now-upated description field at the top of this PR

This PR is an RFC -- if we need another `install_rpath` line for the core library, that's feedback I'd like to get from you or @ihnorton

**Comment by johnkerl** [comment]
I believed I had clean R CI last night.

My apologies; back into draft for the moment.

**Comment by johnkerl** [comment]
(And I certainly had green Ubuntu CI before my rebase, so I'll start by fixing my rebase.)

**Comment by eddelbuettel** [comment]
> I believed I had clean R CI last night.

I believe you had (accidentally ?) turned macOS off.

**Comment by johnkerl** [comment]
@eddelbuettel yes and I believed I had turned it on last thing last night.

It was late, and I may have been mistaken.

I'll fix my recent issues and circle back.

Sorry to disrupt.

**Comment by eddelbuettel** [comment]
No worries!  It is frustrating when the tools get in the way and I really appreciate you are digging at this and have some hopeful smoke signals from the local deployment.

**Comment by johnkerl** [comment]
@nguyenv @eddelbuettel after rebasing on #1301 here is the error I'm now seeing in CI
https://github.com/single-cell-data/TileDB-SOMA/actions/runs/4831511064/jobs/8609059765?pr=1302
which I can also reproduce on an EC2 Linux instance.

Recall from the description section of this PR that the distillation of the build validation, with the changes on this PR, is simply:

```
scripts/bld --prefix=/usr/local/tiledbsoma
cd apis/r
R CMD build --no-build-vignettes --no-manual .
# Also R CMD check ... and R CMD INSTALL ... but this R CMD build step is where we see fails
```

As noted at https://github.com/single-cell-data/TileDB-SOMA/pull/1301#issuecomment-1527776132 we no longer have `span/span.hpp` included in the `libtiledbsoma` install output.

This means that the `R CMD build --no-build-vignettes --no-manual .` promptly fails with

```
g++ -std=gnu++17 -I"/usr/share/R/include" -DNDEBUG -I. -I../inst/include/ -I/usr/local/tiledbsoma/include -I../inst/tiledb/include -I -DR_BUILD  -Iinclude -I'/usr/lib/R/site-library/Rcpp/include' -I'/usr/lib/R/site-library/RcppSpdlog/include'     -fpic  -Wno-deprecated-declarations -Wno-deprecated -c RcppExports.cpp -o RcppExports.o
In file included from /usr/local/tiledbsoma/include/cli/arrow_adapter.h:5,
                 from /usr/local/tiledbsoma/include/tiledbsoma/tiledbsoma:45,
                 from ../inst/include/tiledbsoma_types.h:7,
                 from RcppExports.cpp:4:
/usr/local/tiledbsoma/include/cli/../utils/column_buffer.h:38:10: fatal error: span/span.hpp: No such file or directory
   38 | #include <span/span.hpp>
      |          ^~~~~~~~~~~~~~~
```

which is entirely reasonable since `span/span.hpp` is quite simply not available.

**Comment by johnkerl** [comment]
@eddelbuettel now I'm getting an R CI fail on Ubuntu.

Can you take a look?

**Comment by eddelbuettel** [comment]
> Can you take a look?

Hm.  Build no longer likes or finds the library:

```
libtiledbsoma.so: cannot open shared object file: No such file or directory
```

Not sure why.

**Comment by eddelbuettel** [comment]
I took a bit of a (in hindsight: pointless as failed) detour trying the 'let me compile a quick cmdline program in CI' (which failed over tiledb core not being installed right / pkg-config failing) but with that removed I seem to now be back in good graces.  [This run](https://github.com/single-cell-data/TileDB-SOMA/actions/runs/4834868712/jobs/8616585405) just passed tests.  It still fails in coverage but it did that earlier and I am just going to 'park' that too.  

The macOS side still has blood in the water: 

```
ld: warning: object file (RcppExports.o) was built for newer macOS version (12.0) than being linked (10.14)
ld: warning: object file (rinterface.o) was built for newer macOS version (12.0) than being linked (10.14)
ld: warning: object file (nanoarrow.o) was built for newer macOS version (12.0) than being linked (10.14)
ld: warning: object file (riterator.o) was built for newer macOS version (12.0) than being linked (10.14)
ld: warning: object file (rutilities.o) was built for newer macOS version (12.0) than being linked (10.14)
ld: warning: object file (stats.o) was built for newer macOS version (12.0) than being linked (10.14)
ld: warning: object file (version.o) was built for newer macOS version (12.0) than being linked (10.14)
ld: warning: dylib (/Library/Frameworks/R.framework/R) was built for newer macOS version (11.0) than being linked (10.14)
ld: warning: dylib (/usr/local/lib/libtiledbsoma.dylib) was built for newer macOS version (12.6) than being linked (10.14)
if [ "" != "Windows_NT" ] && [ `uname -s` = 'Darwin' ] && [ -f ../inst/tiledb/lib/libtiledb.dylib ] && [ -f tiledbsoma.so ]; then install_name_tool -change libz.1.dylib @rpath/libz.1.dylib ../inst/tiledb/lib/libtiledb.dylib; install_name_tool -add_rpath @loader_path/../tiledb/lib tiledbsoma.so; fi
installing to /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/Rtmp6QC0jY/Rinst536f281a3550/00LOCK-tiledbsoma/00new/tiledbsoma/libs
```

May be missing a flag as we set for the R package.

**Comment by eddelbuettel** [comment]
So I am back to where I was a few days ago.  'ubuntu-latest' is green, 'macos-latest' not so much. See [this run](https://github.com/single-cell-data/TileDB-SOMA/actions/runs/4834953439/jobs/8616757090).

**Comment by eddelbuettel** [comment]
In the tiledb-r package we do this for macOS on x86_64 (it's `autconf` but you get the idea):

```
## Take care of 10.14 requirement for Intel macOS
if test x"${uname}" = x"Darwin" -a x"${machine}" = x"x86_64"; then
    AC_MSG_CHECKING([for Darwin x86_64 use minimum version override])
    CXX17_MACOS="-mmacosx-version-min=10.14"
    AC_MSG_RESULT([${CXX17_MACOS}])
fi
```

We may have to add the `-mmacosx-version-min=10.14` to CMake _if_ on macOS _and_ using x86_64.

**Comment by johnkerl** [comment]
> So I am back to where I was a few days ago. 'ubuntu-latest' is green, 'macos-latest' not so much.

I did have a successful Mac build before #1301. I think we can get back to success.

**Comment by johnkerl** [comment]
I've tried this six ways:

* My main Mac which has core pre-installed (important developer workflow)
* My personal Mac which does not have core pre-installed (normal user workflow)
* An EC2 Ubuntu instance which has core pre-installed (important developer workflow)
* An EC2 Ubuntu instance which does not have core pre-installed (normal user workflow)
* Linux CI on this repo
* Mac CI on this repo

All are workinig with my most recent commit, except for my main Mac where I am seeing what (I believe is an) unrelated segfault.

**Comment by eddelbuettel** [comment]
This currently runs only Linux

image

as the macOS line is commented-out:

https://github.com/single-cell-data/TileDB-SOMA/blob/440bd4651bc20cda5852c86e05575e517764068d/.github/workflows/r-ci.yml#L21

**Comment by eddelbuettel** [comment]
Looks like the line

https://github.com/single-cell-data/TileDB-SOMA/blob/cc5cb716ed8b38c3346ca6838558a94bc740a8a5/apis/r/src/Makevars.in#L28

was the winner!   Very nice as that uncorks us for now.  

Longer-term we need to adjust this to possible artifact downloads as we do for the tiledb build on macOS.

**Comment by johnkerl** [comment]
Thanks @eddelbuettel -- OK, now even more green. 😎

File: .github/workflows/r-ci.yml [hunk 1]
```diff
@@ -17,8 +17,9 @@ jobs:
     strategy:
       matrix:
         include:
-          - { os: macOS-latest }
+          # TEMP: get Ubuntu working first
```
Comment (eddelbuettel) [inline .github/workflows/r-ci.yml:20]: We can now rempev this line.

File: apis/r/src/Makevars.in [hunk 2]
```diff
@@ -1,15 +1,29 @@
...
-        # if we are
+        # If we are:
         #  - not on Window NT (a tip from data.table)
         #  - on macOS aka Darwin which needs this
         #  - the library is present (implying non-system library use)
         # then let us call install_name_tool
-	if [ "$(OS)" != "Windows_NT" ] && [ `uname -s` = 'Darwin' ] && [ -f ../inst/tiledb/lib/libtiledb.dylib ] && [ -f tiledbsoma.so ]; then install_name_tool -change libz.1.dylib @rpath/libz.1.dylib ../inst/tiledb/lib/libtiledb.dylib; install_name_tool -add_rpath @loader_path/../tiledb/lib tiledbsoma.so; fi
+	#
+	# Pro-tip: here you can run things like `otool -L tiledbsoma.so`, `ls ../inst`, etc. but
+	# the output will only be visible if the R build fails.
```
Comment (eddelbuettel) [inline apis/r/src/Makevars.in:19]: I don't think that is entirely correct.  These line will alway run when the `if` tests above are true.  That includes CI for macOS, but also matching macOS runs outside CI.

File: libtiledbsoma/cmake/Modules/FindTileDB_EP.cmake [hunk 3]
```diff
@@ -132,7 +132,8 @@ else()
   endif()
 endif()
 
-if (EP_TILEDB_BUILT AND TARGET TileDB::tiledb_shared AND NOT TILEDBSOMA_BUILD_R)
+#if (EP_TILEDB_BUILT AND TARGET TileDB::tiledb_shared AND NOT TILEDBSOMA_BUILD_R)
```
Comment (eddelbuettel) [inline libtiledbsoma/cmake/Modules/FindTileDB_EP.cmake:135]: We can remove the line now.

File: libtiledbsoma/cmake/Superbuild.cmake [hunk 4]
```diff
@@ -67,7 +67,7 @@ set(INHERITED_CMAKE_ARGS
   -DSANITIZER=${SANITIZER}
   -DENABLE_ARROW_EXPORT=${ENABLE_ARROW_EXPORT}
   -DOVERRIDE_INSTALL_PREFIX=${OVERRIDE_INSTALL_PREFIX}
-  -DTILEDBSOMA_BUILD_R=${TILEDBSOMA_BUILD_R}
+  #-DTILEDBSOMA_BUILD_R=${TILEDBSOMA_BUILD_R}
```
Comment (eddelbuettel) [inline libtiledbsoma/cmake/Superbuild.cmake:70]: We can remove the line now.

File: apis/r/src/Makevars.in (hunk 2 above)
Comment (johnkerl) [reply apis/r/src/Makevars.in]: @eddelbuettel I mean the output being visible at the terminal when run interactively.

**Pull request merged**